src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,573 @@
+    return std::nullopt;
+  }
+
+  const auto& predicate_expression = std::static_pointer_cast<AbstractPredicateExpression>(predicate_node->predicate());
+
+  // Joins only support these six binary predicates. We rely on PredicateSplitUpRule having split up ANDed chains of
+  // such predicates previously, so that we can process them separately.
+  auto cond_type = predicate_expression->predicate_condition;
+  if (cond_type != PredicateCondition::Equals && cond_type != PredicateCondition::NotEquals &&
+      cond_type != PredicateCondition::LessThan && cond_type != PredicateCondition::LessThanEquals &&
+      cond_type != PredicateCondition::GreaterThan && cond_type != PredicateCondition::GreaterThanEquals) {
+    return std::nullopt;
+  }
+
+  // We can currently only pull equals predicates above aggregate nodes. The other predicate types could be supported
+  // but require more sophisticated reformulations.
+  if (is_below_aggregate && cond_type != PredicateCondition::Equals) {
+    return std::nullopt;
+  }
+
+  // Check that one side of the expression is a correlated parameter and the other a column expression of the LQP below
+  // the predicate node (required for turning it into a join predicate).
+  const auto& binary_predicate_expression = std::static_pointer_cast<BinaryPredicateExpression>(predicate_expression);
+  const auto& left_side = binary_predicate_expression->left_operand();
+  const auto& right_side = binary_predicate_expression->right_operand();
+  PredicateInfo info;
+  info.condition = cond_type;
+  ParameterID parameter_id;
+  if (left_side->type == ExpressionType::CorrelatedParameter) {
+    parameter_id = std::static_pointer_cast<CorrelatedParameterExpression>(left_side)->parameter_id;
+    info.right_operand = right_side;
+  } else if (right_side->type == ExpressionType::CorrelatedParameter) {
+    info.condition = flip_predicate_condition(info.condition);
+    parameter_id = std::static_pointer_cast<CorrelatedParameterExpression>(right_side)->parameter_id;
+    info.right_operand = left_side;
+  } else {
+    return std::nullopt;
+  }
+
+  // We can only use predicates in joins where both operands are columns
+  if (!predicate_node->find_column_id(*info.right_operand)) {
+    return std::nullopt;
+  }
+
+  // Is the parameter one we are concerned with? This catches correlated parameters of outer subqueries and
+  // placeholders in prepared statements.
+  auto expression_it = parameter_mapping.find(parameter_id);
+  if (expression_it == parameter_mapping.end()) {
+    return std::nullopt;
+  }
+
+  info.left_operand = expression_it->second;
+  return info;
+}
+
+/**
+ * Finds predicate nodes to pull up to become join predicates and other nodes that need to be removed or patched.
+ *
+ * This selects all predicates that can be safely pulled up and turned into join predicates. It also collects all
+ * projection nodes above the selected predicates. These need to be removed to make sure that the required columns are
```

Comment:
Right now re remove the projection nodes entirely and rely on `ColumnPruningRule` to put sensible projections back in for us

Reply:
that justification reads a bit odd. Are you aware that the ProjectionNode might be used to calculate expressions (such as `a + b - c`) for subsequent scans/aggregates/joins/sorts?

Reply:
For some reason we didn't consider this at all. In this case we need to handle projection nodes more like aggregate nodes and only add the columns we need. This might actually be a chance to clean up our LQP manipulation (by having less separate logic for alias, aggregate and projection nodes) and handling the diamond LQP case.